### PR TITLE
Automatically start app using `start` script and inject secrets into environment

### PR DESCRIPTION
**Pull request by theneva**
This adds a new default behaviour: If `CMD` is left out of the `Dockerfile`, we inject any available FIAAS secrets into the environment and run the `start` script from `package.json`.

See the changes in README.md for more information!

Output from running a simple test image (which prints nothing else) with different configurations:

```json
---
docker build -t img .

Sending build context to Docker daemon  43.52kB

Step 1/2 : FROM containers.schibsted.io/finntech/node:onbuild-10.9.0
# Executing 5 build triggers
 ---> Using cache
 ---> Using cache
 ---> Using cache
 ---> Using cache
 ---> Using cache
 ---> 4d3e7ef1a2eb
Step 2/2 : COPY index.js .
 ---> Using cache
 ---> ffcd192e930f
Successfully built ffcd192e930f
Successfully tagged img:latest
---

---
docker run img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:37+0000",
  "level": "INFO",
  "message": "FIAAS_ENVIRONMENT is unset, not looking for secrets"
}
---

---
docker run -e FIAAS_ENVIRONMENT=test img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:38+0000",
  "level": "INFO",
  "message": "Secrets directory '/var/run/secrets/fiaas' does not exist, not looking for secrets"
}
---

---
docker run -e FIAAS_ENVIRONMENT=test -v /tmp/empty:/var/run/secrets/fiaas img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:40+0000",
  "level": "INFO",
  "message": "FIAAS_ENVIRONMENT is set to 'test', looking for secrets in '/var/run/secrets/fiaas'"
}
{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:40+0000",
  "level": "INFO",
  "message": "Found no secrets in '/var/run/secrets/fiaas'"
}
---

---
docker run -e FIAAS_ENVIRONMENT=test -v /tmp/secrets:/var/run/secrets/fiaas img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:41+0000",
  "level": "INFO",
  "message": "FIAAS_ENVIRONMENT is set to 'test', looking for secrets in '/var/run/secrets/fiaas'"
}
{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:41+0000",
  "level": "INFO",
  "message": "Found 1 secret(s) in '/var/run/secrets/fiaas'"
}
{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:41+0000",
  "level": "INFO",
  "message": "Exporting secret 'asecret' as 'SECRET_ASECRET'"
}
---

---
docker run -e FIAAS_ENVIRONMENT=test -e SECRETS_DIR=/var/run/secrets/somethingelse img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:45+0000",
  "level": "INFO",
  "message": "Secrets directory '/var/run/secrets/somethingelse' does not exist, not looking for secrets"
}
---

---
docker run -e FIAAS_ENVIRONMENT=test -e SECRETS_DIR=/var/run/secrets/somethingelse -v /tmp/empty:/var/run/secrets/somethingelse img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:46+0000",
  "level": "INFO",
  "message": "FIAAS_ENVIRONMENT is set to 'test', looking for secrets in '/var/run/secrets/somethingelse'"
}
{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:46+0000",
  "level": "INFO",
  "message": "Found no secrets in '/var/run/secrets/somethingelse'"
}
---

---
docker run -e FIAAS_ENVIRONMENT=test -e SECRETS_DIR=/var/run/secrets/somethingelse -v /tmp/secrets:/var/run/secrets/somethingelse img

{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:48+0000",
  "level": "INFO",
  "message": "FIAAS_ENVIRONMENT is set to 'test', looking for secrets in '/var/run/secrets/somethingelse'"
}
{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:48+0000",
  "level": "INFO",
  "message": "Found 1 secret(s) in '/var/run/secrets/somethingelse'"
}
{
  "@version": 1,
  "@timestamp": "2018-08-21T09:39:48+0000",
  "level": "INFO",
  "message": "Exporting secret 'asecret' as 'SECRET_ASECRET'"
}
---
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -63,6 +64,39 @@ This binds the port (3000) inside the container to port 3030 on your Docker host
 
 The application is now available at `http://localhost:3030/`!
 
+## Custom behavior
+
+As mentioned under __Why__, these images are based on the official Docker images for `node:alpine`, and provide all of the same features. However, some behaviors are overridden:
+
+### `CMD`
+
+The default `CMD` behaves exactly as `npm run start`.
```
Comment (trygve-lie) [inline README.md]: `npm start` should be enough.

File: scripts/run-app.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    export "$secret_name=$secret"
+  done
+}
+
+start_app() {
+  # npm provides binaries on PATH in scripts, so we have to do the same
+  export PATH=$PATH:$(pwd)/node_modules/.bin
+
+  local start_script=$(node -p "require('./package.json').scripts.start")
```
Comment (trygve-lie) [inline scripts/run-app.sh]: Why not just `npm start`?

Comment (theneva) [reply scripts/run-app.sh]: we remove `npm` after installing the packages for a smaller image size 😄

File: README.md (hunk 1 above)
Comment (theneva) [reply README.md]: I agree. Fixed!

File: scripts/run-app.sh (hunk 2 above)
Comment (trygve-lie) [reply scripts/run-app.sh]: Is it really that much to save on doing that?

Comment (theneva) [reply scripts/run-app.sh]: /shrug. I don't want to change it for this feature :)

Comment (SimenB) [reply scripts/run-app.sh]: npm/yarn is over 20% of the image

File: README.md [hunk 3]
```diff
@@ -11,6 +11,7 @@ These images exist to simplify development of Node applications at FINN. They ar
 - Automatically install dependencies using `yarn` or `npm`, depending on which lockfile is present
 - Provide ready-to-use `onbuild` versions which handle most of what you'll need for the image to be built (see below)
 - Install [`dumb-init`](https://github.com/Yelp/dumb-init) which fixes some issues with signal forwarding
+- Expose FIAAS secrets directly in the environment (more below)
```
Comment (SimenB) [inline README.md]: what's fiaas?

Comment (SimenB) [reply README.md]: (this image is open source)

File: scripts/run-app.sh (hunk 2 above)
Comment (trygve-lie) [reply scripts/run-app.sh]: I have zero issues with removing `yarn`, but to be honest I do have doubts about removing `npm`. `npm` is shipped with node and is a part of node. Due to that I would actually expect `npm` to be pressent in the runtime for the whole life time of the process.

File: README.md [hunk 4]
```diff
@@ -63,6 +64,39 @@ This binds the port (3000) inside the container to port 3030 on your Docker host
 
 The application is now available at `http://localhost:3030/`!
 
+## Custom behavior
+
+As mentioned under __Why__, these images are based on the official Docker images for `node:alpine`, and provide all of the same features. However, some behaviors are overridden:
```
Comment (SimenB) [inline README.md]: link to `why` instead if emphasis

File: scripts/run-app.sh [hunk 5]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env sh
+
+log() {
```
Comment (SimenB) [inline scripts/run-app.sh]: nice!

File: README.md (hunk 4 above)
Comment (theneva) [reply README.md]: Done 😄

**Comment by theneva** [comment]
I feel ready to squash & merge this. One last look @SimenB? ❤️

**Comment by theneva** [comment]
We should also look at releasing new Node 8 images with this change. Is that just `./release.sh 8.11.4-alpine` to overwrite the existing images?

File: scripts/run-app.sh (hunk 2 above)
Comment (theneva) [reply scripts/run-app.sh]: I suppose this is relevant for the cases where people do "npm run <something>" as part of their scripts, but that doesn't really change with this PR (and isn't very useful when you're just starting the app -- it's much more common for linting and building). Maybe we should support "prestart" though?

**Comment by SimenB** [comment]
`./release 8.11.4-1`

File: scripts/run-app.sh (hunk 2 above)
Comment (SimenB) [reply scripts/run-app.sh]: nah, I don't think it's needed.

**Pull request merged**